File: packages/molweave_core/src/molweave_core/selection.py
from __future__ import annotations

from collections.abc import Iterable, Mapping
from math import dist
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from molweave_core.molecular import Atom, Chain, NormalizedStructureV1, Residue

SelectionGranularity = Literal["atom", "residue", "chain", "structure"]
SelectionSource = Literal["viewer", "project", "sequence", "inspector", "saved"]
SelectionMode = Literal["replace", "add", "subtract"]
SelectionPredicateField = Literal[
    "atom_name",
    "atom_index",
    "atom_reference",
    "element",
    "residue_name",
    "residue_number",
    "chain",
    "structure",
]


class AtomReference(BaseModel):
    model_config = ConfigDict(frozen=True)

    structure_id: str = Field(min_length=1)
    atom_id: int = Field(ge=1)


class SelectionV1(BaseModel):
    model_config = ConfigDict(frozen=True)

    schema_version: Literal[1] = 1
    atoms: list[AtomReference] = Field(default_factory=list)
    granularity: SelectionGranularity = "atom"
    source: SelectionSource = "inspector"

    @model_validator(mode="after")
    def canonical_atoms(self) -> SelectionV1:
        keys = [(reference.structure_id, reference.atom_id) for reference in self.atoms]
        if keys != sorted(set(keys)):
            raise ValueError("Selection atom references must be unique and canonically ordered")
        return self


def selection(
    references: Iterable[AtomReference],
    *,
    granularity: SelectionGranularity = "atom",
    source: SelectionSource = "inspector",
) -> SelectionV1:
    unique = {(reference.structure_id, reference.atom_id): reference for reference in references}
    return SelectionV1(
        atoms=[unique[key] for key in sorted(unique)],
        granularity=granularity,
        source=source,
    )


def combine(
    current: SelectionV1,
    operand: SelectionV1,
    mode: SelectionMode,
) -> SelectionV1:
    current_keys = {(item.structure_id, item.atom_id) for item in current.atoms}
    operand_keys = {(item.structure_id, item.atom_id) for item in operand.atoms}
    if mode == "replace":
        result = operand_keys
    elif mode == "add":
        result = current_keys | operand_keys
    else:
        result = current_keys - operand_keys
    return selection(
        (
            AtomReference(structure_id=structure_id, atom_id=atom_id)
            for structure_id, atom_id in result
        ),
        granularity=operand.granularity,
        source=operand.source,
    )


def invert(
    current: SelectionV1,
    structures: Mapping[str, NormalizedStructureV1],
    *,
    source: SelectionSource = "inspector",
) -> SelectionV1:
    universe = selection(
        (
            AtomReference(structure_id=structure_id, atom_id=atom.id)
            for structure_id, structure in structures.items()
            for atom in structure.atoms
        ),
        source=source,
    )
    return combine(universe, current, "subtract")


def expand(
    current: SelectionV1,
    structures: Mapping[str, NormalizedStructureV1],
    granularity: SelectionGranularity,
    *,
    source: SelectionSource = "inspector",
) -> SelectionV1:
    selected = {(item.structure_id, item.atom_id) for item in current.atoms}
    expanded: list[AtomReference] = []
    for structure_id, structure in structures.items():
        selected_atoms = {
            atom_id
            for selected_structure, atom_id in selected
            if selected_structure == structure_id
        }
        if not selected_atoms:
            continue
        atoms_by_id = {atom.id: atom for atom in structure.atoms}
        residue_ids = {
            atom.residue_id
            for atom_id in selected_atoms
            if (atom := atoms_by_id.get(atom_id)) is not None and atom.residue_id is not None
        }
        residue_by_id = {residue.id: residue for residue in structure.residues}
        chain_ids = {
            residue_by_id[residue_id].chain_id
            for residue_id in residue_ids
            if residue_id in residue_by_id
        }
        for atom in structure.atoms:
            include = (
                atom.id in selected_atoms
                if granularity == "atom"
                else atom.residue_id in residue_ids
                if granularity == "residue"
                else (
                    atom.residue_id is not None
                    and atom.residue_id in residue_by_id
                    and residue_by_id[atom.residue_id].chain_id in chain_ids
                )
                if granularity == "chain"
                else True
            )
            if include:
                expanded.append(AtomReference(structure_id=structure_id, atom_id=atom.id))
    return selection(expanded, granularity=granularity, source=source)


def select_by_predicate(
    structures: Mapping[str, NormalizedStructureV1],
    field: SelectionPredicateField,
    value: str,
    *,
    source: SelectionSource = "inspector",
) -> SelectionV1:
    expected = value.strip().casefold()
    matches: list[AtomReference] = []
    for structure_id, structure in structures.items():
        residues = {residue.id: residue for residue in structure.residues}
        chains = {chain.id: chain for chain in structure.chains}
        for atom in structure.atoms:
            residue = residues.get(atom.residue_id) if atom.residue_id is not None else None
            chain = chains.get(residue.chain_id) if residue is not None else None
            actual = _predicate_value(field, structure_id, atom, residue, chain)
            if actual is not None and actual.casefold() == expected:
                matches.append(AtomReference(structure_id=structure_id, atom_id=atom.id))
    return selection(matches, source=source)


def within_distance(
    seed: SelectionV1,
    structures: Mapping[str, NormalizedStructureV1],
    distance_angstrom: float,
    *,
    granularity: Literal["atom", "residue"] = "atom",
    source: SelectionSource = "inspector",
) -> SelectionV1:
    if distance_angstrom < 0:
        raise ValueError("Distance must be non-negative")
    seed_coordinates = [
        atom.coordinates
        for reference in seed.atoms
        if (structure := structures.get(reference.structure_id)) is not None
        if (atom := _atom(structure, reference.atom_id)) is not None
    ]
    matched: list[AtomReference] = []
    if seed_coordinates:
        for structure_id, structure in structures.items():
            for atom in structure.atoms:
                if any(
                    dist(atom.coordinates, coordinate) <= distance_angstrom
                    for coordinate in seed_coordinates
                ):
                    matched.append(AtomReference(structure_id=structure_id, atom_id=atom.id))
    result = selection(matched, source=source)
    return (
        expand(result, structures, "residue", source=source) if granularity == "residue" else result
    )


def valid_references(
    current: SelectionV1,
    valid_atom_ids: Mapping[str, set[int]],
    *,
    source: SelectionSource | None = None,
) -> tuple[SelectionV1, int]:
    kept = [
        reference
        for reference in current.atoms
        if reference.atom_id in valid_atom_ids.get(reference.structure_id, set())
    ]
    return (
        selection(
            kept,
            granularity=current.granularity,
            source=source or current.source,
        ),
        len(current.atoms) - len(kept),
    )


def _atom(structure: NormalizedStructureV1, atom_id: int) -> Atom | None:
    return next((atom for atom in structure.atoms if atom.id == atom_id), None)


def _predicate_value(
    field: SelectionPredicateField,
    structure_id: str,
    atom: Atom,
    residue: Residue | None,
    chain: Chain | None,
) -> str | None:
    if field == "atom_name":
        return atom.name
    if field == "atom_index":
        return str(atom.id)
    if field == "atom_reference":
        return f"{structure_id}:{atom.id}"
    if field == "element":
        return atom.element
    if field == "structure":
        return structure_id
    if residue is None:
        return None
    if field == "residue_name":
        return residue.name
    if field == "residue_number":
        number = residue.author_number or residue.label_number
        return str(number) if number is not None else None
    return chain.name if chain is not None else None

File: apps/web/src/selection/selection.ts
import type {
  AtomReference,
  NormalizedStructure,
  Selection,
  SelectionGranularity,
  SelectionMode,
  SelectionSource,
} from "../api/types";

export type StructureMap = ReadonlyMap<string, NormalizedStructure>;
export type PredicateField =
  | "atom_name"
  | "atom_index"
  | "atom_reference"
  | "element"
  | "residue_name"
  | "residue_number"
  | "chain"
  | "structure";

const key = (reference: AtomReference) =>
  `${reference.structure_id}\u0000${String(reference.atom_id).padStart(12, "0")}`;

export function canonicalSelection(
  atoms: Iterable<AtomReference>,
  granularity: SelectionGranularity = "atom",
  source: SelectionSource = "inspector",
): Selection {
  const unique = new Map<string, AtomReference>();
  for (const reference of atoms) unique.set(key(reference), reference);
  return {
    schema_version: 1,
    atoms: [...unique.entries()]
      .sort(([left], [right]) => left.localeCompare(right))
      .map(([, reference]) => reference),
    granularity,
    source,
  };
}

export const emptySelection = (): Selection =>
  canonicalSelection([], "atom", "inspector");

export function combineSelection(
  current: Selection,
  operand: Selection,
  mode: SelectionMode,
): Selection {
  if (mode === "replace") return canonicalSelection(operand.atoms, operand.granularity, operand.source);
  const result = new Map(current.atoms.map((reference) => [key(reference), reference]));
  for (const reference of operand.atoms) {
    if (mode === "add") result.set(key(reference), reference);
    else result.delete(key(reference));
  }
  return canonicalSelection(result.values(), operand.granularity, operand.source);
}

export function selectionMode(event: Pick<MouseEvent, "altKey" | "ctrlKey" | "metaKey" | "shiftKey">): SelectionMode {
  if (event.altKey) return "subtract";
  if (event.ctrlKey || event.metaKey || event.shiftKey) return "add";
  return "replace";
}

export function structureSelection(
  structureId: string,
  structure: NormalizedStructure,
  source: SelectionSource,
): Selection {
  return canonicalSelection(
    structure.atoms.map((atom) => ({ structure_id: structureId, atom_id: atom.id })),
    "structure",
    source,
  );
}

export function residueSelection(
  structureId: string,
  structure: NormalizedStructure,
  residueId: number,
  source: SelectionSource,
): Selection {
  return canonicalSelection(
    structure.atoms
      .filter((atom) => atom.residue_id === residueId)
      .map((atom) => ({ structure_id: structureId, atom_id: atom.id })),
    "residue",
    source,
  );
}

export function chainSelection(
  structureId: string,
  structure: NormalizedStructure,
  chainId: number,
  source: SelectionSource,
): Selection {
  const residueIds = new Set(
    structure.residues.filter((residue) => residue.chain_id === chainId).map((residue) => residue.id),
  );
  return canonicalSelection(
    structure.atoms
      .filter((atom) => atom.residue_id !== null && residueIds.has(atom.residue_id))
      .map((atom) => ({ structure_id: structureId, atom_id: atom.id })),
    "chain",
    source,
  );
}

export function invertSelection(
  current: Selection,
  structures: StructureMap,
): Selection {
  const universe = canonicalSelection(
    [...structures].flatMap(([structureId, structure]) =>
      structure.atoms.map((atom) => ({ structure_id: structureId, atom_id: atom.id })),
    ),
    "atom",
    "inspector",
  );
  return combineSelection(universe, current, "subtract");
}

export function expandSelection(
  current: Selection,
  structures: StructureMap,
  granularity: SelectionGranularity,
): Selection {
  const expanded: AtomReference[] = [];
  for (const [structureId, structure] of structures) {
    const selectedIds = new Set(
      current.atoms
        .filter((reference) => reference.structure_id === structureId)
        .map((reference) => reference.atom_id),
    );
    if (selectedIds.size === 0) continue;
    const atomsById = new Map(structure.atoms.map((atom) => [atom.id, atom]));
    const residueIds = new Set(
      [...selectedIds]
        .map((atomId) => atomsById.get(atomId)?.residue_id)
        .filter((residueId): residueId is number => residueId !== null && residueId !== undefined),
    );
    const residuesById = new Map(structure.residues.map((residue) => [residue.id, residue]));
    const chainIds = new Set(
      [...residueIds]
        .map((residueId) => residuesById.get(residueId)?.chain_id)
        .filter((chainId): chainId is number => chainId !== undefined),
    );
    for (const atom of structure.atoms) {
      const include =
        granularity === "atom"
          ? selectedIds.has(atom.id)
          : granularity === "residue"
            ? atom.residue_id !== null && residueIds.has(atom.residue_id)
            : granularity === "chain"
              ? atom.residue_id !== null &&
                chainIds.has(residuesById.get(atom.residue_id)?.chain_id ?? -1)
              : true;
      if (include) expanded.push({ structure_id: structureId, atom_id: atom.id });
    }
  }
  return canonicalSelection(expanded, granularity, "inspector");
}

export function predicateSelection(
  structures: StructureMap,
  field: PredicateField,
  rawValue: string,
): Selection {
  const expected = rawValue.trim().toLocaleLowerCase();
  const matches: AtomReference[] = [];
  for (const [structureId, structure] of structures) {
    const residues = new Map(structure.residues.map((residue) => [residue.id, residue]));
    const chains = new Map(structure.chains.map((chain) => [chain.id, chain]));
    for (const atom of structure.atoms) {
      const residue = atom.residue_id === null ? undefined : residues.get(atom.residue_id);
      const chain = residue ? chains.get(residue.chain_id) : undefined;
      const value =
        field === "atom_name"
          ? atom.name
          : field === "atom_index"
            ? String(atom.id)
            : field === "atom_reference"
              ? `${structureId}:${atom.id}`
              : field === "element"
            ? atom.element
            : field === "residue_name"
              ? residue?.name
              : field === "residue_number"
                ? String(residue?.author_number ?? residue?.label_number ?? "")
                : field === "chain"
                  ? chain?.name
                  : structureId;
      if (value?.toLocaleLowerCase() === expected) {
        matches.push({ structure_id: structureId, atom_id: atom.id });
      }
    }
  }
  return canonicalSelection(matches, "atom", "inspector");
}

export function selectedEntryIds(current: Selection): Set<string> {
  return new Set(current.atoms.map((reference) => reference.structure_id));
}

export function selectionSummary(current: Selection, structures: StructureMap) {
  const residues = new Set<string>();
  const chains = new Set<string>();
  for (const reference of current.atoms) {
    const structure = structures.get(reference.structure_id);
    const atom = structure?.atoms.find((item) => item.id === reference.atom_id);
    const residue = structure?.residues.find((item) => item.id === atom?.residue_id);
    if (residue) {
      residues.add(`${reference.structure_id}:${residue.id}`);
      chains.add(`${reference.structure_id}:${residue.chain_id}`);
    }
  }
  return {
    atoms: current.atoms.length,
    residues: residues.size,
    chains: chains.size,
    structures: selectedEntryIds(current).size,
  };
}
